content/content/Advance Usage/custom plugins.md

```diff
@@ -0,0 +1,130 @@
+---
+title: "Custom Plugins"
+date: 2021-11-16T12:38:44-05:00
+draft: false
+---
+
+### How to develop a plugin binary
+
+This document covers how to write a plugin binary using crane-lib. It requires:
+1. Go development environment setup
+2. Optionally, an overview of the crane toolkit
+ 
+A binary plugin for crane-lib is meant to be a simple Go program in the following format:
+ 
+```
+package main
+
+import (
+	"fmt"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/konveyor/crane-lib/transform"
+	"github.com/konveyor/crane-lib/transform/cli"
+)
+
+func main() {
+	fields := []transform.OptionalFields{
+		{
+			FlagName: "my-flag",
+			Help:     "What the flag does",
+			Example:  "true",
+		},
+	}
+	cli.RunAndExit(cli.NewCustomPlugin("MyCustomPlugin", "v1", fields, Run))
+}
+
+func Run(request transform.PluginRequest) (transform.PluginResponse, error) {
+	// plugin writers need to write custome code here.
+	resp := transform.PluginResponse{}
+	// prepare the response
+	return resp, nil
+}
+```
+
+All it does is read an input from stdin, calls the Run function with the input object passed as unstructured and prints the return value of `Run` function on stdout.
+
+The json passed in via stdin is a `transform.PluginRequest` which consists of an inline unstructured object and an optional `Extras` map containing additional flags. Without any `Extras` the format is identical to the json output from a `kubectl get -o json` call. When adding extra params, a map field "extras" is added at the top level (parallel to "apiVersion", "kind", etc.).
+
+During the development of the plugin, one can iterate by passing in the JSON object on stdin manually. For example, if the above code is compiled and run with the following json as input, the output will be `{}`.
```

Comment:
I'm wondering whether we should make sure that the example plugin above sets the plugin response version to `transform.V1` like we do with the openshift plugin. We'd keep `IsWhiteOut` and `Patches` empty, though. If that's done, then the output will be `{"version":"v1"}` instead of `{}`